### PR TITLE
BOOKKEEPER-896: VM-local transport

### PR DESCRIPTION
**Pull request by eolivelli**

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java [hunk 1]
```diff
@@ -76,14 +81,20 @@
...
         InetSocketAddress bindAddress;
         if (conf.getListeningInterface() == null) {
             // listen on all interfaces
             bindAddress = new InetSocketAddress(conf.getBookiePort());
         } else {
-            bindAddress = Bookie.getBookieAddress(conf).getSocketAddress();
+            bindAddress = bookieAddress.getSocketAddress();
         }
-        listenOn(bindAddress);
+        listenOn(bindAddress,bookieAddress);
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java]: ",bookieAddress" -> ", bookieAddress"

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/LocalBookiesRegistry.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.bookkeeper.net.BookieSocketAddress;
+
+/**
+ * Local registry for embedded Bookies
+ */
+class LocalBookiesRegistry {
+    
+    private final static ConcurrentHashMap<BookieSocketAddress,Boolean> localBookiesRegistry = new ConcurrentHashMap<>();
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/LocalBookiesRegistry.java]: too long line?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 3]
```diff
@@ -195,8 +204,11 @@ private void connect() {
         bootstrap.setOption("child.receiveBufferSize", conf.getClientReceiveBufferSize());
         bootstrap.setOption("writeBufferLowWaterMark", conf.getClientWriteBufferLowWaterMark());
         bootstrap.setOption("writeBufferHighWaterMark", conf.getClientWriteBufferHighWaterMark());
-
-        ChannelFuture future = bootstrap.connect(addr.getSocketAddress());
+        SocketAddress bookieAddr = addr.getSocketAddress();        
+        if (channelFactory instanceof LocalClientChannelFactory) {
+            bookieAddr = new LocalAddress(addr.getSocketAddress().toString());
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: How about adding a method #getLocalAddress in BookieSocketAddress?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/LocalBookKeeperTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
+package org.apache.bookkeeper.client;
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/LocalBookKeeperTest.java]: move the package line down to the file comment?

**Comment by sijie** [comment]
the patch looks good to me. several minor comments.

**Comment by sijie** [comment]
the change looks good to me. +1.

**Comment by sijie** [comment]
@eolivelli the branch is not mergeable with current master. could you help rebasing your branch to latest master?

**Comment by eolivelli** [comment]
@sijie rebased

**Comment by merlimat** [comment]
:+1:

**Pull request closed without merging**